pkg/kubelet/util/pluginwatcher/plugin_watcher.go

```diff
@@ -176,65 +188,85 @@ func (w *Watcher) handleFsNotifyEvent(event fsnotify.Event) error {
 		return fmt.Errorf("stat file %s failed: %v", event.Name, err)
 	}
 
-	if !fi.IsDir() {
+	switch mode := fi.Mode(); {
+	case mode&os.ModeSocket != 0:
 		return w.registerPlugin(event.Name)
+	case mode.IsDir():
+		if err := w.traversePluginDir(event.Name); err != nil {
+			return fmt.Errorf("failed to traverse plugin path %s, err: %v", event.Name, err)
+		}
+	default:
+		glog.V(2).Infof("ignoring %s", event.Name)
 	}
-
-	if err := w.traversePluginDir(event.Name); err != nil {
-		return fmt.Errorf("failed to traverse plugin path %s, err: %v", event.Name, err)
-	}
-
 	return nil
 }
 
 // Start watches for the creation of plugin sockets at the path
-func (w *Watcher) Start() error {
+func (w *Watcher) Start() (<-chan error, error) {
 	glog.V(2).Infof("Plugin Watcher Start at %s", w.path)
 	w.stopCh = make(chan interface{})
 
 	// Creating the directory to be watched if it doesn't exist yet,
 	// and walks through the directory to discover the existing plugins.
 	if err := w.init(); err != nil {
-		return err
+		return nil, err
 	}
 
 	fsWatcher, err := fsnotify.NewWatcher()
 	if err != nil {
-		return fmt.Errorf("failed to start plugin fsWatcher, err: %v", err)
+		return nil, fmt.Errorf("failed to start plugin fsWatcher, err: %v", err)
 	}
 	w.fsWatcher = fsWatcher
 
 	if err := w.traversePluginDir(w.path); err != nil {
 		fsWatcher.Close()
-		return fmt.Errorf("failed to traverse plugin socket path, err: %v", err)
+		return nil, fmt.Errorf("failed to traverse plugin socket path, err: %v", err)
 	}
 
+	w.errCh = make(chan error)
 	w.wg.Add(1)
 	go func(fsWatcher *fsnotify.Watcher) {
 		defer w.wg.Done()
 		for {
 			select {
 			case event := <-fsWatcher.Events:
-				//TODO: Handle errors by taking corrective measures
 				go func() {
 					err := w.handleFsNotifyEvent(event)
 					if err != nil {
-						glog.Errorf("error %v when handle event: %s", err, event)
+						atomic.AddInt64(&w.eventHandleErrCount, 1)
+						glog.Errorf("handle event: %s failed, %v", event, err)
 					}
 				}()
 				continue
 			case err := <-fsWatcher.Errors:
-				if err != nil {
-					glog.Errorf("fsWatcher received error: %v", err)
+				if err == fsnotify.ErrEventOverflow {
+					select {
+					case w.errCh <- err:
+					case <-w.stopCh:
+						return
+					}
 				}
+				glog.Errorf("fsWatcher received error: %v", err)
 				continue
 			case <-w.stopCh:
 				fsWatcher.Close()
 				return
+			case <-time.After(eventHandleErrCountingInterval * time.Second):
+				errRate := atomic.LoadInt64(&w.eventHandleErrCount) / eventHandleErrCountingInterval
+				if errRate >= eventHandleErrThreshold {
```

Comment:
@hui could you explain the purpose of having eventHandleErrThreshold and eventHandleErrCountingInterval timeout? What if we don't have them?

Reply:
@jiayingz thanks, i should add a comment to code, btw, my github id is "figo".

The timer here is used to calculate error rate, it will post one error to error channel only when error rate passed threshold, the rationale behind is:
we only need to glog for single error, it becomes major issue when error happening too often (it indicates some plugins have problem, the new added test is simulating this behavior).

Reply:
Is the main purpose of this logic to prevent DOS from a single bad plugin? I do feel it is important to have some kind of rate limiting for that purpose, but we probably need to track registration and error rates at PluginInfo.name level for that purpose. To limit the scope of this PR, can we leave this logic out  of this PR for now? Please feel free to create separate issue and PR if you are interested in helping implement that feature, which I personally feel is quite important. Thanks!

Reply:
issue filed: https://github.com/kubernetes/kubernetes/issues/67517
your suggestion sounds good to me, i will keep the fsnotify overflow error logic here, and move the error rate part out, thanks

Reply:
@jiayingz  change updated, PTAL